hedged.go

```diff
@@ -53,40 +53,49 @@ func (ht *hedgedTransport) RoundTrip(req *http.Request) (*http.Response, error)
 	defer cancel()
 	req = req.WithContext(ctx)
 
-	var res interface{}
-	resultCh := make(chan interface{}, ht.upto)
+	var res *http.Response
+	var err error
+	resultCh := make(chan *http.Response, ht.upto)
+	errorCh := make(chan error, ht.upto)
 
+Loop:
 	for sent := 0; ; sent++ {
 		if sent < ht.upto {
 			runInPool(func() {
 				resp, err := ht.rt.RoundTrip(req)
 				if err != nil {
-					resultCh <- err
+					errorCh <- err
 				} else {
 					resultCh <- resp
 				}
 			})
 		}
+		// try to read result channel first, before blocking on all other channels
+		select {
+		case res = <-resultCh:
+			break Loop
+		default:
+		}
 
 		select {
 		case res = <-resultCh:
+			break Loop
+		case err = <-errorCh:
+			continue
 		case <-ctx.Done():
-			res = ctx.Err()
+			err = ctx.Err()
 		case <-time.After(ht.timeout):
```

Comment:
During #1 re-understanding I've realised that we should create `time.Ticker` manually and stop it after the loop, but that's out of scope for this PR, just mentioning for the future.